### ISSUE TITLE
Prefix global variables in themes

### ISSUE THREAD

**Issue by grappler** [WordPress/WordPress-Coding-Standards#1043]
I was running PHPCS on `_s` with the following in in the ruleset

```xml
<rule ref="WordPress.NamingConventions.PrefixAllGlobals">
	<properties>
		<property name="prefixes" type="array" value="_s" />
	</properties>
</rule>
```

The problem was that I was getting false positives. For example the variables used in the theme templates were being triggered. These variables do not need to be prefixed as they are within a template.

Also `$GLOBALS['content_width'] = apply_filters( '_s_content_width', 640 );` was marked as an issue.

The only valid error was `$themecolors` in `_s/inc/wpcom.php` which will be whitelisted as it is wp.com specific.

**Comment by GaryJones** [comment]
> These variables do not need to be prefixed as they are within a template.

They are more than likely variables in the global scope though, and may interact with plugins that also fiddle with variables in the global scope.

**Comment by grappler** [comment]
> They are more than likely variables in the global scope though, and may interact with plugins that also fiddle with variables in the global scope.

Sorry, I don't fully understand what you mean.

Here is a code example.
`$archive_content = '<p>' . sprintf( esc_html__( 'Try looking in the monthly archives. %1$s', '_s' ), convert_smilies( ':)' ) ) . '</p>';`
https://github.com/Automattic/_s/blob/master/404.php#L47

**Comment by GaryJones** [comment]
Right - that's still an unprefixed global variable. Changing the variable name to `$_s_archive_content` should be sufficient.

**Comment by grappler** [comment]
@GaryJones I don't believe that it is global variable. The TRT has never required variables with a template to be prefixed. I am happy to be proven wrong. How would one overwrite the variable? Something like a proof of concept.

**Comment by jrfnl** [comment]
> Also $GLOBALS['content_width'] = apply_filters( '_s_content_width', 640 ); was marked as an issue.

I'm pretty sure I looked at allowing `$content_width` and a few other typical ones to be overwritten as that's what they are intended for, but not sure what happened with that.
Might have been left for the "the list of globals needs to be updated" issue #924.

> For example the variables used in the theme templates were being triggered. These variables do not need to be prefixed as they are within a template.

If I remember correctly, template files are normally included & executed from within a function call within WP, so all variables within them would be considered local to that function call.
However, that is no guarantee that those files will *always* be called in that way, so IMHO, prefixing variables within template files would still be considered best practice.

**Comment by dingo-d** [comment]
>The TRT has never required variables with a template to be prefixed.

I was under the impression that only variables (properties) and methods within classes are allowed to be unprefixed. Everything else should be prefixed...

**Comment by grappler** [comment]
> If I remember correctly, template files are normally included & executed from within a function call within WP, so all variables within them would be considered local to that function call.

That is my understanding too.

> However, that is no guarantee that those files will always be called in that way, so IMHO, prefixing variables within template files would still be considered best practice.

> However, that is no guarantee that those files will always be called in that way

With WordPress's track record with backwards compatibility I don't expect it changing soon. 99% of themes do not prefix their variables in templates. 

> I was under the impression that only variables (properties) and methods within classes are allowed to be unprefixed. Everything else should be prefixed...

Any variables in the global scope need to be prefixed. If variables in a template need to prefixed or not for the TRT has been discussed in the #themereview channel in the past.

Your themes does not prefix the variables e.g. https://themes.svn.wordpress.org/expire/1.0.9/footer.php

**Comment by jrfnl** [comment]
> I was under the impression that only variables (properties) and methods within classes are allowed to be unprefixed. Everything else should be prefixed...

Only variables which are within a **_local_** scope - in contrast to the _global_ scope are allowed to be unprefixed. Local scopes are classes, traits, methods, functions, closures.

If a file is included by PHP from within a function call and the file declares variables, they will be imported to the local scope of the function, not to the global scope. This is how PHP works.

So considering that WP includes template files from within a function call, yes, they would be considered _local_ but only _if included via WP in the normal manner_.

> >    However, that is no guarantee that those files will always be called in that way
>
> With WordPress's track record with backwards compatibility I don't expect it changing soon. 

I did not mean just by WP. Templates files can also be included by plugins or even by a separate app which creates a custom front-end for WP.

**Comment by grappler** [comment]
> Templates files can also be included by plugins or even by a separate app which creates a custom front-end for WP.

True, I suppose theme developers could exclude those templates that are being used by WP from this sniff.

I suppose what remains is to have a whitelist of global variables that are not prefixed.

**Comment by jrfnl** [comment]
The "non-prefixed global variables in themes" issue has been tackled in the `WPThemeReview` standard in https://github.com/WPTRT/WPThemeReview/pull/208.

As for the issue with `$content_width`, PR #1773 should take care of that.
